cpp/src/arrow/scalar_test.cc

```diff
@@ -902,6 +1214,37 @@ TEST(TestDictionaryScalar, Basics) {
   }
 }
 
+TEST(TestDictionaryScalar, ValidateErrors) {
+  auto index_ty = int16();
+  auto value_ty = utf8();
+  auto dict = ArrayFromJSON(value_ty, R"(["alpha", "beta", "gamma"])");
```

Comment:
For completeness, would you add a case where a valid index encodes a null dictionary entry?
```suggestion
  auto dict = ArrayFromJSON(value_ty, R"(["alpha", "beta", "gamma", null])");
```

Reply:
In this case, should `DictionaryScalar::is_valid` be true or false?

Reply:
I'd say `is_valid==true`